app/src/main/kotlin/at/bitfire/davdroid/sync/TaskSyncer.kt

```diff
@@ -11,109 +11,92 @@ import android.content.SyncResult
-            if (Build.VERSION.SDK_INT >= 26) {
-                /* Warning: If setAccountVisibility is called, Android 12 broadcasts the
-                   AccountManager.LOGIN_ACCOUNTS_CHANGED_ACTION Intent. This cancels running syncs
-                   and starts them again! So make sure setAccountVisibility is only called when necessary. */
-                val am = AccountManager.get(context)
-                if (am.getAccountVisibility(account, providerName.packageName) != AccountManager.VISIBILITY_VISIBLE)
-                    am.setAccountVisibility(account, providerName.packageName, AccountManager.VISIBILITY_VISIBLE)
-            }
-
-            val accountSettings = accountSettingsFactory.forAccount(account)
-
-            // 1. find task collections to be synced
-            val remoteCollections = mutableMapOf<HttpUrl, Collection>()
-            val service = db.serviceDao().getByAccountAndType(account.name, Service.TYPE_CALDAV)
-            if (service != null)
-                for (collection in db.collectionDao().getSyncTaskLists(service.id))
-                    remoteCollections[collection.url] = collection
+    @AssistedFactory
+    interface Factory {
+        fun create(account: Account, authority: String, extras: Array<String>, syncResult: SyncResult): TaskSyncer
+    }
 
-            // 2. delete/update local task lists and determine new remote collections
-            val updateColors = accountSettings.getManageCalendarColors()
-            val newCollections = HashMap(remoteCollections)
-            for (list in DmfsTaskList.find(account, taskProvider, LocalTaskList.Factory, null, null))
-                list.syncId?.let {
-                    val url = it.toHttpUrl()
-                    val info = remoteCollections[url]
-                    if (info == null) {
-                        logger.fine("Deleting obsolete local task list $url")
-                        list.delete()
-                    } else {
-                        // remote CollectionInfo found for this local collection, update data
-                        logger.log(Level.FINE, "Updating local task list $url", info)
-                        list.update(info, updateColors)
-                        // we already have a local task list for this remote collection, don't create a new local task list
-                        newCollections -= url
-                    }
-                }
+    private lateinit var taskProvider: TaskProvider
 
-            // 3. create new local task lists
-            for ((_,info) in newCollections) {
-                logger.log(Level.INFO, "Adding local task list", info)
-                LocalTaskList.create(account, taskProvider, info)
-            }
+    override val serviceType: String
+        get() = Service.TYPE_CALDAV
+    override fun localCollections(provider: ContentProviderClient): List<LocalTaskList>
+        = DmfsTaskList.find(account, taskProvider, LocalTaskList.Factory, null, null)
```

Comment:
TaskSyncer still needs to acquire its own TaskProvider, since finding the local sync enabled collections with `DmfsTaskList.find()` requires it.It would require some more rewriting in ical4android for us to be able to do that.